tests/views/test_collection_instrument_view.py

```diff
@@ -0,0 +1,127 @@
+import base64
+
+from application.controllers.collection_instrument import UPLOAD_SUCCESSFUL, COLLECTION_EXERCISE_NOT_FOUND, \
+    INVALID_CLASSIFIER
+from application.controllers.session_decorator import with_db_session
+from application.models.models import ExerciseModel, InstrumentModel, BusinessModel, SurveyModel
+from flask import current_app
+from six import BytesIO
+from tests.test_client import TestClient
+from unittest.mock import patch
+
+
+class TestCollectionInstrumentView(TestClient):
+    """ Collection Instrument view unit tests"""
+
+    def test_collection_instrument_upload(self):
+
+        # Given an upload file and a patched survey_id response
+        survey_id = {"surveyId": "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87"}
+        data = dict(file=(BytesIO(b'test data'), 'test.xls'))
+        with patch('application.controllers.collection_instrument.service_request', return_value=survey_id):
+
+            # When a post is made to the upload end point
+            response = self.client.post(
+                '/collection-instrument-api/1.0.2/upload/{ref}/{file}'
+                .format(ref='cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87', file='test.xls'),
+                headers=self.get_auth_headers(),
+                data=data,
+                content_type='multipart/form-data')
+
+            # Then the file uploads successfully
+            self.assertStatus(response, 200)
+            self.assertEquals(response.data.decode(), UPLOAD_SUCCESSFUL)
+
+    def test_download_exercise_csv(self):
+
+        # Given a patched exercise
+        instrument = InstrumentModel(data='test_data', length=999)
+        exercise = ExerciseModel(exercise_id='cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87')
+        business = BusinessModel(ru_ref='test_ru_ref')
+        instrument.exercises.append(exercise)
+        instrument.businesses.append(business)
+
+        with patch('application.controllers.collection_instrument.query_exercise_by_id', return_value=exercise):
```

Comment:
Would there be value in actually querying a database even if it's sqlite? It would confirm we are using the right filters/methods via slqalchemy? (We may already be exercising it but have missed it)

Reply:
I think a bigger piece needs to be done with integration tests as well, will raise a card.

Reply:
Speak with Rich Weeks. I think he spotted an issue with the drivers that SQL alchemy uses and was getting different results when using sqlite vs postgres. This could make test results misleading.